File: packages/react-native-sdk/src/streaming-edit-batcher.ts
export interface StreamingEditUpdate {
  id: number;
  content: string;
  editedAt: string;
  streaming?: boolean;
}

export interface StreamingEditBatcherOptions {
  intervalMs?: number;
  maxWaitMs?: number;
}

export function createStreamingEditBatcher(
  apply: (updates: StreamingEditUpdate[]) => void,
  options: StreamingEditBatcherOptions = {},
) {
  const intervalMs = options.intervalMs ?? 80;
  const maxWaitMs = options.maxWaitMs ?? 200;
  const pending = new Map<number, StreamingEditUpdate>();
  let timer: ReturnType<typeof setTimeout> | null = null;
  let firstQueuedAt = 0;

  function flush() {
    if (timer) { clearTimeout(timer); timer = null; }
    if (pending.size === 0) { firstQueuedAt = 0; return; }
    const batch = Array.from(pending.values());
    pending.clear();
    firstQueuedAt = 0;
    apply(batch);
  }

  function schedule() {
    const now = Date.now();
    if (!firstQueuedAt) firstQueuedAt = now;
    const elapsed = now - firstQueuedAt;
    const delay = elapsed >= maxWaitMs ? 0 : intervalMs;
    if (timer) clearTimeout(timer);
    timer = setTimeout(flush, delay);
  }

  return {
    push(update: StreamingEditUpdate) { pending.set(update.id, update); schedule(); },
    flush,
  };
}

File: packages/flutter-sdk/lib/src/streaming_edit_batcher.dart
import 'dart:async';

class StreamingEditBatcher {
  final void Function(List<Map<String, dynamic>> updates) apply;
  final int intervalMs;
  final int maxWaitMs;
  final Map<int, Map<String, dynamic>> _pending = {};
  Timer? _timer;
  int _firstQueuedAt = 0;

  StreamingEditBatcher(this.apply, {this.intervalMs = 80, this.maxWaitMs = 200});

  void push(Map<String, dynamic> update) {
    final id = update['id'] as int;
    _pending[id] = update;
    _schedule();
  }

  void flush() {
    _timer?.cancel();
    _timer = null;
    if (_pending.isEmpty) { _firstQueuedAt = 0; return; }
    final batch = _pending.values.toList();
    _pending.clear();
    _firstQueuedAt = 0;
    apply(batch);
  }

  void _schedule() {
    final now = DateTime.now().millisecondsSinceEpoch;
    if (_firstQueuedAt == 0) _firstQueuedAt = now;
    final elapsed = now - _firstQueuedAt;
    final delay = elapsed >= maxWaitMs ? 0 : intervalMs;
    _timer?.cancel();
    _timer = Timer(Duration(milliseconds: delay), flush);
  }

  void dispose() { _timer?.cancel(); }
}
